### DISCUSSION TITLE
Letters showcase animation

### DISCUSSION THREAD

**Discussion by bensapirstein** [coldtype/coldtype#135] [Q&A]
How can I achieve this exponential ease-in-out letter animations in Coldtype? I recently created an animation that showcased letters from different scripts using this effect, but I would like to replicate it using Coldtype. Additionally, I want to avoid limiting the animation to one letter at a time, as I did by looping over the letters for each transition. Instead, I would like to create a smoother transition with a small overlap between when a letter eases out and the next one starts to ease in.

Bonus question: How can I expand this animation to create a matrix comparison between different fonts? I envision each row displaying the same letter in various fonts, and as the animation progresses, each row would shift upwards to the row above while the row below eases in.

https://github.com/coldtype/coldtype/assets/14977606/0d7e0c7c-f154-4ace-9daa-c4a2cff02d7c

**Comment by bensapirstein** [comment]
I have successfully implemented an animation that showcases Phoenician letters. However, I am encountering difficulties in properly implementing the exp ease-in-out effect for these letters. In my previous animation, I manually implemented the ease-in-out interpolation using a low p parameter of 0.2, following the approach described in this resource: [https://www.s-ings.com/scratchpad/exponential-easing/](url).

Currently, the Coldtype animation uses a linear interpolation, and I am facing the issue of no overlapping between the entry and exit of the letters. To provide a better understanding of my desired outcome, here is an example of a similar animation created by a talented friend of mine, which could serve as a good reference: [https://twigl.app/?ss=-NWs_iq8FutNBDHgwXkp](url).

Ideally, I envision the animation to consist of multiple rows, each representing the same letter in different fonts (or writing systems). These letters should transition from right-to-left and bottom-to-top. The purpose of this animation is to showcase a database I am utilizing for an academic project.

I would greatly appreciate any guidance and suggestions on how I can implement this vision successfully. Your assistance would be invaluable to me. 🙏

```python
from coldtype import *
from coldtype.fx.skia import phototype

fnt_ph = Font.Cacheable("~/Fonts/semiticRegular/NotoSansPhoenician-Regular.ttf")
fnt_lt = "assets/ColdtypeObviously-VF.ttf"

# Define the string of Phoenician letters
letters = "𐤀𐤁𐤂𐤃𐤄𐤅𐤆𐤇𐤈𐤉𐤊𐤋𐤌𐤍𐤎𐤏𐤐𐤑𐤒𐤓𐤔𐤕"
titles = ["ʾālep", "bēt", "gīml", "dālet", "he", "wāw", "zayin", "ḥēt", "ṭēt", "yod", "kāp", "lāmed", "mēm", "nūn", "śāmek", "ʿayin", "pē", "ṣādē", "qōp", "reš", "šīn", "tāw"]
N = len(letters) 
fps = 60 # frames per second
fpl = 2 * fps # frames per letter

@animation(timeline=(fpl*len(letters), fps), bg=0)
def latin_letters(f):
    # Calculate the index of the current frame
    index = f.i // fpl

    # Calculate the interpolation time relative to the current letter
    t = f.i / fpl - index

    return (StSt(letters[index],fnt_ph,800)
            .align(f.a.r)
            # translate so that slides from the right
            .translate((0.5 - t) * (1080+800), -80)
            .f(1)
            .ch(phototype(f.a.r, blur=3, cut=90, cutw=30, fill=bw(1))),
            (StSt(titles[index],fnt_lt,80))
            .align(f.a.r)
            .translate(0, 450)
            .f(1))
```

https://github.com/coldtype/coldtype/assets/14977606/782603a2-6efa-4267-b075-462aa629c0fe

**Comment by stenson** [comment]
Given your code above I think the easiest thing would be to apply a custom easing curve to your variable `t`, using `coldtype.timing.easing.ease` and a custom `P` curve. It's very tweakable (as noted in the code) to get a specific look to the easing curve. I also added a line to show the curve visually, but obviously you'd want to comment that out once you've got it dialed in.

```python
from coldtype import *
from coldtype.fx.skia import phototype
from coldtype.timing.easing import ease

fnt_ph = Font.Find("NotoSansPhoenician-Regular.ttf")
fnt_lt = "assets/ColdtypeObviously-VF.ttf"

# Define the string of Phoenician letters
letters = "𐤀𐤁𐤂𐤃𐤄𐤅𐤆𐤇𐤈𐤉𐤊𐤋𐤌𐤍𐤎𐤏𐤐𐤑𐤒𐤓𐤔𐤕"
titles = ["ʾālep", "bēt", "gīml", "dālet", "he", "wāw", "zayin", "ḥēt", "ṭēt", "yod", "kāp", "lāmed", "mēm", "nūn", "śāmek", "ʿayin", "pē", "ṣādē", "qōp", "reš", "šīn", "tāw"]
N = len(letters) 
fps = 60 # frames per second
fpl = 2 * fps # frames per letter

@animation(timeline=(fpl*len(letters), fps), bg=0)
def latin_letters(f):
    # Calculate the index of the current frame
    index = f.i // fpl

    ease_curve = P().withRect(1000, lambda r, p: p
        .moveTo(r.psw)
        .ioEaseCurveTo(r.pc, 21, 30) # 21 and 50 are arbitrary
        .ioEaseCurveTo(r.pne, 5, 30)) # 5 and 30 are arbitrary

    # Calculate the interpolation time relative to the current letter
    t = f.i / fpl - index
    t = ease(ease_curve, t)[0] # apply custom easing from visual curve

    return (P(
        ease_curve.copy().fssw(-1, 0.5, 4), # visual easing curve
        StSt(letters[index],fnt_ph,800)
            .align(f.a.r)
            # translate so that slides from the right
            .translate((0.5 - t) * (1080+800), -80)
            .f(1),
        StSt(titles[index], Font.RecMono(), 80)
            .align(f.a.r)
            .translate(0, 450)
            .f(1)))
```

**Reply by stenson** [reply]
Came with a slightly different solution and added it as an example to the repo: https://github.com/coldtype/coldtype/blob/main/examples/animations/letters_easing.py

**Reply by bensapirstein** [reply]
Thanks for this example! I'll look into it and learn from it some good Coldtype practices.

I came up with an alternative solution by adding a new entry of a custom-made Easing to the easing.eases dictionary. This particular easing closely resembles the rate of eeio, and I find it to be a better fit for my needs:

```python
from easing_functions.easing import EasingBase
import coldtype.timing.easing as easing

class MyExponentialEaseInOut(EasingBase):
    def func(self, t, p=0.3):
        if t == 0 or t == 1:
            return t

        if t < 0.5:
            return 0.5 * math.pow(t * 2, p)
        return 1 - 0.5 * math.pow(2 - t * 2, p)
    
easing.eases["meeio"] = MyExponentialEaseInOut
[...]
    l = at.current(0).now()
    e1 = l.e("meeio", 0)
```
